src/vma/lwip/tcp_in.c

```diff
@@ -599,6 +599,16 @@ tcp_process(struct tcp_pcb *pcb, tcp_in_data* in_data)
       tcp_rst(in_data->ackno, in_data->seqno + in_data->tcplen,
         in_data->tcphdr->dest, in_data->tcphdr->src, pcb);
     }
+    /* received SYN? SYN_RCVD via SYN */
+    else if (in_data->flags & TCP_SYN) {
+      err = tcp_enqueue_flags(pcb, TCP_SYN);
+      if (err == ERR_OK) {
+        /* Changed the pcbs state now */
+        set_tcp_state(pcb, SYN_RCVD);
+        pcb->in_flag_hndlg |= TCP_SYN;
+        TCP_SET_ERROR_STATUS(pcb, 0);
+      }
+    }
```

Comment:
1. Is this addition supposed to handle this case ?
https://tools.ietf.org/html/rfc793#section-3.4 
The "three-way handshake" is the procedure used to establish a
  connection.  This procedure normally is initiated by one TCP and
  responded to by another TCP.  The procedure also works if two TCP
  simultaneously initiate the procedure.  When simultaneous attempt
  occurs, each TCP receives a "SYN" segment which carries no
  acknowledgment after it has sent a "SYN". 

2. tcp_enqueue_flags is used to enqueue TCP options for transmission but i dont see any tcp_out() call here ?
Where we send the data ?

3. set_tcp_state(pcb, SYN_RCVD) is called only from tcp_listen_input() while with this change it can also be called from tcp_process during connect() - is this valid ?

Reply:
1. It seems yes. 2. The fast path is used. Review -wip. 3. tcp connect during connect  review - wip.